data/blog/enabling-out-of-the-box-performance-insights-in-the-unity-sdk.md

```diff
@@ -0,0 +1,188 @@
+
+## Understanding the Unity Lifecycle: Finding Key Points for Instrumentation
+
+The game works in a super tight loop, typically updating anywhere from 30 to 60 times per second but the sky is the limit. Creating a span to measure every single tick is not feasible. We needed to look at some overarching actions like some set of logical operations we would want to capture.
+
+![Unity Lifecycle](/images/autoinstrumentation-in-unity-via-ilweaving/unity-life-cycle.png)
+
+### The Challenge of Defining Transactions and Spans
+
+For measuring how long something takes Sentry has two working concepts: Transactions and Spans. Transactions are single [instances of an activity or a service](https://docs.sentry.io/product/performance/transaction-summary/#what-is-a-transaction), like loading of a page or some async task. Spans are individual measurements that are nested within a transaction. Conceptionally, we're trying to find places to start and stop a big stopwatch for bigger, and very specific actions that we want to measure. And then we are looking for sup-tasks within that action that we could capture with smaller stopwatches. But how does a transaction fit within the the frame of a game? What instance of a service, that is already built into the engine, could a transaction represent?
+
+For all its features Unity is still a blank canvas for you to create any kind of game. That means there are, other than the general lifecycle, not very many fixed points that the SDK could hook into to start and stop a span. There are a whole bunch of one-time events like button clicks but how would the SDK hook into whatever happens behind the button click? How would the SDK know when to finish the span?
+
+## Universal Events in Unity: Startup and Scene Loading
+
+All games need to startup and the startup procedure is the same for all Unity games and consists of loading of systems, the splash-screen (if applicable) and the loading of the initial scene. Scene loading in general is another great fixed point. Everything within Unity exists within the context of a scene. Some games load them additionally, some games swap them, some games only ever have one. But at least that one gets loaded during startup.
+
+With this we had our transaction hooks. We know when the startup is starting and finished. And we can hook into the scene manager to time scene load and scene load finish events.
+
+## Adding Granularity: Populating Transactions with Spans
+
+Now that we have our overarching operation that we’re trying to time we’re now looking for smaller actions that happen within. Looking towards Unity’s lifecycle helps us out once more. The initialization happens for every GameObject during its creation or, if it is an initial part of the scene, during the scene’s loading. For all GameObjects the one method that gets invoked is the `Awake` call. But that’s user’s code. How would the SDK instrument non-SDK code without asking to user to do it for us?
```

Comment:
Would be nice to mention HTTP requests here too.

"Other integrations the SDK provide, such as HttpClient, also enter this transaction. If those HTTP requests happen during the scene load, they might be useful context to why a scene loads slowly"

Or something

But also:

```suggestion
Now that we have our overarching operation that we’re trying to time we’re now looking for smaller actions that happen within. Looking towards Unity’s lifecycle helps us out once more. The initialization happens for every GameObject during its creation or, if it is an initial part of the scene, during the scene’s loading. For all GameObjects the one method that gets invoked is the `Awake` call. And that’s user’s code, exactly what you'd want to know how its performing since you have the most control over improving. But how would the SDK instrument non-SDK code without asking to user to do it for us?
```